seqr/views/apis/anvil_workspace_api.py

```diff
@@ -1,48 +1,117 @@
 """APIs for management of projects related to AnVIL workspaces."""
 
 import logging
-from django.contrib.auth.decorators import login_required
+import json
 
+from django.shortcuts import redirect
+from django.contrib.sites.shortcuts import get_current_site
+from django.core.exceptions import PermissionDenied
+
+from seqr.models import Project, CAN_EDIT
+from seqr.views.utils.json_to_orm_utils import create_model_from_json
 from seqr.views.utils.json_utils import create_json_response
-from settings import API_LOGIN_REQUIRED_URL
+from seqr.views.utils.file_utils import load_uploaded_file
+from seqr.views.utils.terra_api_utils import add_service_account
+from seqr.views.utils.pedigree_info_utils import parse_pedigree_table
+from seqr.views.apis.individual_api import add_or_update_individuals_and_families
+from seqr.utils.communication_utils import send_load_data_email
+from seqr.views.utils.permissions_utils import google_auth_required, check_workspace_perm, has_workspace_perm
 
 logger = logging.getLogger(__name__)
 
 
-@login_required(login_url=API_LOGIN_REQUIRED_URL)
+def add_individuals_and_families(project, individual_records, user):
+    _, updated_individuals = add_or_update_individuals_and_families(
+        project, individual_records=individual_records, user=user
+    )
+
+    return '\n'.join(['Individual ID'] + [individual.individual_id for individual in updated_individuals])
+
+
+@google_auth_required
 def anvil_workspace_page(request, namespace, name):
     """
-    Redirect to the loading data from workspace page or redirect to the project if the project exists.
+    Validate the workspace and project before loading data.
+
+    :param request: Django request object.
+    :param namespace: The namespace (or the billing account) of the workspace.
+    :param name: The name of the workspace. It also be used as the project name.
+    :return Redirect to a page depending on if the workspace permissions or project exists.
 
-    :param request:
-    :param namespace:
-    :param name:
-    :return:
     """
-    # To be implemented
-    return create_json_response({})
+    if not has_workspace_perm(request.user, CAN_EDIT, namespace, name, can_share=True):
+        message = "{user} does not have sufficient permissions for workspace {namespace}/{name}".format(
+            user=request.user, namespace=namespace, name=name)
+        logger.warning(message)
+        raise PermissionDenied(message)
```

Comment:
Instead of calling `check_workspace_perm()` I changed it to call `has_workspace_perm()` so that I can add logging before raising the exception. Or, do you want the logging to be added in `check_workspace_perm()`?

Reply:
AsI said before, the logging should be in `check_workspace_perm`

Reply:
Okay